Create reference documentation for current state of package management support

Leonidas-from-XIV opened the pull request:
For a proper release it makes sense to document the package management extensions to the `dune-project` and `dune-workspace`. This PR documents the current state, though it is to be discussed if this is what we plan to roll out as final form of the feature.

Leonidas-from-XIV committed:
Create reference documentation for current package management support

Signed-off-by: Marek Kubica <marek@tarides.com>

Leonidas-from-XIV commented:
@rgrinberg While writing the documentation I had some observations:

  * Both `pin` stanzas use `url` while `repository` uses `source`. I think it would be nicer to just always refer to URLs via `(url ..)`.
  * `(package (name <foo>))` is cumbersome. How about adding a shorthand of `(package <foo>)`? If people want a version they could still use `(name <foo>) (version <bar>)` or even `(package <foo>.<bar>)` using the common OPAM convention, which cuts down on the nesting.
  * `pins` in `dune-workspace` are named `sources` internally, since that's what they were named originally. Maybe worth renaming it internally too, to avoid confusion where the dune-lang calls it something but then the code calls it something else internally.

What do you think?

rgrinberg commented:
> (package (name <foo>)) is cumbersome. How about adding a shorthand of (package <foo>)? If people want a version they could still use (name <foo>) (version <bar>) or even (package <foo>.<bar>) using the common OPAM convention, which cuts down on the nesting.

Sympathetic to the problem, but I'm not entirely convinced. We have a similar shorthand for the `rule` stanza, and my experience is that it hasn't pulled its own weight. One problem with such shorthands, is that they really complicate the life for tooling to be built on top of dune files. With enough such shortcuts, one would not be able to write anything without using a library that understands these tricks.

If there's enough demands from users with some real world examples of long pins, there would be a much stronger case for this.

gridbugs commented on doc/reference/dune-project/pin.rst:
```diff
@@ -0,0 +1,43 @@
+      .. describe:: (version <string>)
+
+         The version that the package should be assumed to be.
+
```
```suggestion
This is optional, and its default value is "dev".
```

gridbugs commented on doc/reference/dune-project/pin.rst:
```diff
@@ -0,0 +1,43 @@
+   .. describe:: (url <string>)
+
+      The URL of the package source.
+
```
```suggestion
This can be a path to a directory on the local file system or remote git repository. Local paths can be absolute or relative, and may optionally begin with `file://` though this is not necessary. Remote git repository urls must begin with `git+https://`, for example `git+https://github.com/user/repo`.
```

gridbugs commented on doc/reference/dune-workspace/context.rst:
```diff
@@ -15,6 +15,11 @@ the description of an opam switch, as follows:
    build's context artifacts will be stored.
 
+-  ``(lock_dir <path>)`` specifies the lock directory that will be used for
+   building this context (if any). See the
```
```suggestion
   building this context. This is optional and its default value is `dune.lock`. See the
```

gridbugs commented on doc/reference/dune-workspace/lock_dir.rst:
```diff
@@ -0,0 +1,72 @@
+   .. describe:: (path <string>)
+
+      The location in the source tree where the lock directory will be
+      created or read from.
```
```suggestion
      created or read from. This is optional and its default value is "dune.lock".
```

gridbugs reviewed (approved):
Looks good. I added suggestions for a couple of clarifications and examples.

maiste commented on doc/reference/dune-project/pin.rst:
```diff
@@ -0,0 +1,43 @@
+
+.. warning::
+
+   Dune Package Management is not final yet and the configuration options
+   are subject to change.
```
:art: I would add a pointer to the explanation about package management you wrote. But this is just my trying to create a graph in the documentation :smile:

maiste commented on doc/reference/dune-project/pin.rst:
```diff
@@ -0,0 +1,43 @@
+
+      The URL of the package source.
+
+      This must be specified.
+
+   .. describe:: (package ...)
+
+      Defines which package is to be pinned.
+
+      This must be specified.
+
+      .. describe:: (name <string>)
+
+         The name of the package.
+
+         This must be specified.
+
+      .. describe:: (version <string>)
+
+         The version that the package should be assumed to be.
```
Do we have a convention for `required` vs `optional` in the documentation? If not, I think we should think about some uniformity (not for this PR but keep it in a corner of our head).

:art:  I would suggest emphasising this, otherwise.
```suggestion
      This must **be specified**.

   .. describe:: (package ...)

      Defines which package is to be pinned.

      This must **be specified**.

      .. describe:: (name <string>)

         The name of the package.

         This must **be specified**.

      .. describe:: (version <string>)

         The version that the package should be assumed to be.
```

Leonidas-from-XIV replied:
> Do we have a convention for required vs optional in the documentation?

I think in most places it only mentions that a field is required and all fields are optional otherwise. (`rg required doc` vs `rg optional doc`) thus for consistency I would just write that fields are required where they are and nothing otherwise.

maiste commented on doc/reference/dune-workspace/lock_dir.rst:
```diff
@@ -0,0 +1,72 @@
+   Dune Package Management is not final yet and the configuration options
+   are subject to change.
+
+This stanza configures settings for lock directories what are relevant to
```
```suggestion
This stanza configures settings for lock directories who are relevant to
```
Otherwise, I'm not sure to get the meaning of the sentence :sweat: 

Shouldn't we mention that this is related to package management (apart from the warning)?

maiste commented on doc/reference/dune-workspace/lock_dir.rst:
Shouldn't this file specified `required` / `be specified` fields?

Leonidas-from-XIV replied:
None of them are required and it seems the documentation points out if fields are required (which sort of makes sense, most fields are optional and many people don't know all the options of all stanzas).

maiste commented on doc/reference/dune-workspace/pin.rst:
```diff
@@ -0,0 +1,46 @@
+
+.. warning::
+
+   Dune Package Management is not final yet and the configuration options
+   are subject to change.
+
+This stanza is used to define additional package sources.
```
:art: I would suggest the same as the other two: link to package management and a description that this is related to package management apart from within the warning.

Again, feel free to ignore if you don't think this is necessary :smile:

Leonidas-from-XIV replied:
I've changed it a bit to mention that :+1:

maiste commented on doc/reference/dune-workspace/pin.rst:
```diff
@@ -0,0 +1,46 @@
+
+.. note::
+
+   Defining a pin does not enable it by default. It needs to be enabled in a
```
Shouldn't it be more explicit that it needs to be in `(pins ...)` (IIUC)?

Leonidas-from-XIV replied:
I think it is fairly obvious, the stanza is not that large so I would defer if people find it unclear.

maiste commented on doc/reference/dune-workspace/pin.rst:
```diff
@@ -0,0 +1,46 @@
+
+         The name of the package.
+
+         This must be specified.
```
```suggestion
         This must **be specified**.
```
:art:  Same question as the other part of the documentation.

maiste commented on doc/reference/dune-workspace/repository.rst:
```diff
@@ -0,0 +1,37 @@
+
+.. warning::
+
+   Dune Package Management is not final yet and the configuration options
+   are subject to change.
+
+This stanza defines a new named repository and attaches a source location to it.
```
:art:  Same as the others: link and explicit about the context?

maiste commented on doc/reference/dune-workspace/repository.rst:
```diff
@@ -0,0 +1,37 @@
+
+.. note::
+
+   Defining a repository does not enable it in project by default. It needs to be
+   enabled in a lock directory using the :doc:`/reference/dune-workspace/lock_dir`
```
I would also point the `(repositories ...)` stanza here to clarify which one we are talking about.

Leonidas-from-XIV replied:
I don't think I can link all that exactly but also `lock_dir` doesn't have that many options :-)

maiste commented on doc/reference/dune-workspace/repository.rst:
```diff
@@ -0,0 +1,37 @@
+
+      The name used to refer to the respository. Names have to be unique.
+
+      This must be specified.
```
:art: 
```suggestion
      This must **be specified**.
```

maiste commented on doc/reference/dune-workspace/repository.rst:
```diff
@@ -0,0 +1,37 @@
+      extensive control of the version by specifying an exact revision, tag or
+      branch.
+
+      This must be specified.
```
:art: 
```suggestion
      This must **be specified.**
```

maiste reviewed (commented):
Many remarks are some duplicates in case you want to change them across the documentation. I have used the :art: to mentioned "cosmetic" changes and in case you want to ignore them.

Leonidas-from-XIV commented:
@gridbugs Thanks, I've incorporated a lot of your changes with some rewording. It's definitely good to mention the default values.

I think I've incorporated most of the comments in a way where it is consistent with the rest of the documentation so unless someone objects I'd merge it (despite the pointless coverage CI failure).
